### PR TITLE
2.x Update function signatures for ::get_post() and ::get_posts()

### PR DESCRIPTION
**Pull request by gchtr**
## Issue

**Ticket**: #2090, #2094 related to `Timber\Post` objects.

### Add `Helper::doing_it_wrong()`

~~I added a `Helper::doing_it_wrong()` function, that can be used in combination with a `@expectedIncorrectUsage` tag in the DocBlock of a test.~~

~~This is helpful for functionality that will not be deprecated, but removed, like certain function parameters in `Timber::get_post()` and `Timber::get_posts()`.~~

Added in https://github.com/timber/timber/pull/2196.

### Remove query methods that won’t work anymore

The following methods for retrieving post were removed. Yes, removed and not deprecated, because there will be no backwards compatibility.

- **Query strings** - Removed the method of getting posts via a query string (like `post_type=portfolio`) and added a doing-it-wrong message.
- **Getting a post by post slug** – Added a doing-it-wrong message to help developers in the transition. Alternatively, we could add a new function `Timber::get_post_from_slug()` that could do this (Discussion in #2073).
- **Getting a post by post name** – This case is catched with the same doing-it-wrong message as getting a post by post slug (Discussion in #2073).

### Deprecate `Timber::query_post()` and `Timber::query_posts()`

The `Timber::query_post()` and `Timber::query_posts()` functions weren’t properly deprecated yet.

### Return a `PostCollection`

As per the new API, we always return a `PostCollection` when using `Timber::get_posts()`. That’s why I changed the current implementation through `PostGetter::get_posts()` to use `true` for the `return_collection` parameter in `Timber::get_posts()`.

However, a lot of tests still break, because `Timber::get_posts()` doesn’t return a `PostCollection()` with the ability to handle pagination yet. I guess we’ll first have to make `Timber::get_posts()` use the new API before this can be merged. @acobster How would you proceed here?

## Impact

This will break backwards compatibility with how `Timber::get_post()` and `Timber::get_posts()` worked before. But I added checks and warnings to help developers in the transition.

## Usage Changes

See #2073.

## Considerations

To keep discussion in one place, I added some remarks and questions in https://github.com/timber/timber/pull/2073#issuecomment-570954735.

## Testing

Yes, but work in progress.

## Todo

- [ ] Define default `$options` for `Timber::get_posts()`
- [ ] Make `Timber::get_posts()` return a `PostCollection` that can handle pagination.
- [ ] Use `Iterable` as return type for `Timber::get_posts()`, see https://github.com/timber/timber/pull/2073#issuecomment-571814739.

### CODE REVIEW

**Comment by gchtr** [comment]
As per https://github.com/timber/timber/pull/2073#issuecomment-571375271, I guess I should change this against the `2.x-factories` branch.

File: lib/Timber.php [hunk 1]
```diff
@@ -124,67 +124,164 @@ class_alias( 'Timber\Timber', 'Timber' );
...
+		if ( func_num_args() > 1 ) {
+			// @todo Add link to Class Maps documentation.
+			Helper::doing_it_wrong(
+				'Timber::get_post()',
+				'The $PostClass parameter for passing in the post class to use in Timber::get_post() was removed in Timber 2.0. Use Class Maps (LINK HERE) instead.',
+				'2.0.0'
+			);
+		}
+
+		// @todo Will this be handled here or inside factories, or somewhere else?
```
Comment (acobster) [inline lib/Timber.php]: For UserFactories we decided the analogous `get_current_user_id` would be in `Timber::get_user()`, so I think we should be consistent and do this here. Definitely not in a factory in any case.

File: lib/Timber.php [hunk 2]
```diff
@@ -124,67 +124,164 @@ class_alias( 'Timber\Timber', 'Timber' );
...
 	/**
 	 * Query post.
 	 *
 	 * @api
-	 * @deprecated since 2.0.0 Use `new Timber\Post()` instead.
+	 * @deprecated since 2.0.0 Use `Timber::get_post()` instead.
 	 *
-	 * @param mixed  $query
-	 * @param string $PostClass
+	 * @param null|int|\WP_Post $post
```
Comment (acobster) [inline lib/Timber.php]: this should change from `null` to `false`

File: lib/Timber.php [hunk 3]
```diff
@@ -124,67 +124,164 @@ class_alias( 'Timber\Timber', 'Timber' );
...
+	 * @param null|int|\WP_Post $post Optional. Post ID or WP_Post object to get a Timber Post
+	 *                                object. Default `null`.
+	 *
+	 * @return \Timber\Post|bool `Timber\Post` object or an instance of a child class of
+	 *                           `Timber\Post` if a post was found, `false` if no post was found.
 	 */
-	public static function get_post( $query = false, $PostClass = 'Timber\Post' ) {
-		return PostGetter::get_post($query, $PostClass);
+	public static function get_post( $post = null ) {
+		if ( is_string( $post ) ) {
```
Comment (acobster) [inline lib/Timber.php]: sometimes we get IDs as strings, e.g. from `$_GET`. Factories support this already; should we support it here? We could add an `!is_numeric()` check to be certain that we definitely don't have an ID. Only consideration there would be if the `post_title` were a purely string...which would be weird and edge-casey. I feel like that's an okay trade-off...?

Comment (gchtr) [reply lib/Timber.php]: Good call! Resolved in e5e0a4543b85b2110f0d69552e6fe6b16537ad4e.

File: lib/Timber.php (hunk 1 above)
Comment (gchtr) [reply lib/Timber.php]: Resolved in f2d754502dd44436cf37993f3d2ff141350759a6.

File: lib/Timber.php (hunk 2 above)
Comment (gchtr) [reply lib/Timber.php]: Resolved in a2abbb38096cd5abc762c4827c37d037b1215a78.

**Comment by gchtr** [comment]
@acobster I guess I should rebase this pull request against `2.x-factories`, too?

**Comment by jarednova** [comment]
@gchtr are you waiting on me for a review here? I noticed the *WIP* tag so I had assumed you were still at-work on this. 

There are a number of [failing tests](https://travis-ci.com/github/timber/timber/jobs/347276809), mainly related to pagination. Do we need to resolve those here, or is that better done once the merge you and @acobster were discussing is complete?

**Comment by acobster** [comment]
@jarednova I'm assuming there's going to be a lot of lifting to get these two methods working as we get them over to Factories. That work should probably still be done in `2.x-factories` IMHO before we bring it into mainline `2.x`.

**Comment by acobster** [comment]
@gchtr @jarednova I just realized I may have misunderstood the question. I don't really have an opinion about whether these fixes happen in `2.x-post-functions` or `2.x-factories`, in case that's what you were asking.

Sorry, my brain has been kinda everywhere recently. Maybe you know the feeling. :upside_down_face:

**Comment by jarednova** [comment]
@gchtr the PR has my 👍  on the substance you raise in the original ticket, but I think we should work out (not necessarily solve, but just we're on the same page on this todo you left ...):

> Make Timber::get_posts() return a PostCollection that can handle pagination

Right now `PostQuery`s are essentially this — a subclass of `PostCollection` that have methods to handle for pagination and the associated data/logic. I think it's crucial that pagination be attached to the returned object as opposed to a separate request. 

I'm imagining that most of this code can be brought over into a new class w/o too much hassle; but just wanted to make sure it wasn't lost.

Another thing I encountered when doing some investigation on my local is just the number of jumps from class-to-class a call to `Timber::get_posts` makes. I'm very much in favor of simplifying this as we adapt to factories even if it means compromising some backwards compatibility in places. The current chain of functions/methods is dizzying (and can't be great for performance either).

You might already be aware or have plans for these. But just wanted to call them out so they don't bite us later!

**Comment by acobster** [comment]
Ported into #2316 in https://github.com/timber/timber/commit/3fabc800f33e85c3e3de15be62d0f43e2f1f2426. Calling this good.

**Pull request closed without merging**